### DISCUSSION TITLE
Unable to load a native DLL

### DISCUSSION THREAD

**Discussion by bernardkaplan** [pythonnet/pythonnet#1985] [Q&A]
### Environment

-   Pythonnet version: 3.9.12
-   Python version: 3.0.0-r1
-   Operating System: Windows 11
-   .NET Runtime: 6.0.110

### Details

The dotnet6 clr fails to find the native dll on which an assembly depends.

Below is a simple example of a project using ParquetSharp and the exception that is raised:

```bash
System.DllNotFoundException: Unable to load DLL 'ParquetSharpNative' or one of its dependencies: Le module spécifié est introuvable. (0x8007007E)
   at ParquetSharp.LogicalType.LogicalType_None(IntPtr& logicalType)
   at ParquetSharp.ExceptionInfo.Return[TValue](GetAction`1 getter)
   at ParquetSharp.LogicalType.None()
   at ParquetSharp.LogicalTypeFactory..cctor()
```

Of course, if I copy the `ParquetSharpNative.dll` file from `runtimes/win-x64/native` next to the `ParquetSharp.dll` assembly, the exception won't be raised.

This work-around isn't very elegant nor satisfying, especially when the code is shipped in a package.

Is there a better way to let the clr know where to find the native dll ?  
Maybe using the Native.deps.json which seems to define the location for each system ?

### Project

```ini
pythonnet_native
|--dotnet
|  |--Native
|     |--Native.csproj
|     |--Reader.cs
|--pythonnet_native.py
```

### C# project

Reader.cs
```csharp
using ParquetSharp;

namespace Native
{
    public class Reader
    {
        public static void Open(string path)
        {
            using var file = new ParquetFileReader(path);
        }
    }
}
```

Native.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ParquetSharp" Version="8.0.0" />
  </ItemGroup>
  
  <Target Name="CopyDlls" AfterTargets="Build">
    <ItemGroup>
      <MySourceFiles Include="$(TargetDir)**/*.*" />
    </ItemGroup>
    <Copy 
      SourceFiles="@(MySourceFiles)" 
      DestinationFolder="$(ProjectDir)../../libs/%(RecursiveDir)"
    />
  </Target>

</Project>
```
### Python code

pythonnet_native.py
```python
from pythonnet import load

load("coreclr")
import os

import clr
import pandas as pd

df = pd.DataFrame({"A": range(5)})
df.to_parquet("A.parquet")
clr.AddReference(os.path.abspath("libs/Native.dll"))

from Native import Reader  # type:ignore

Reader.Open("A.parquet")
```

### Setup

Python environment
```bash
python -m venv venv
.\venv\Scripts\activate
python -m pip install pythonnet pandas fastparquet
```

Build project
```bash
dotnet build .\dotnet\Native\Native.csproj
```

**Comment by filmor** [comment]
Have you tried `dotnet publish`?

**Reply by bernardkaplan** [reply]
My goal is to build a python package that can run on both win-x64 and linux-x64.

When I call `dotnet publish --arch x64 --os win` the ParquetSharpNative.dll is copied next to the ParquetSharp.dll assembly.
If I choose `--os linux` ParquetSharpNative.so will be copied instead. The other assemblies look unchanged.

Does it mean I simply have to ship both ParquetSharpNative.dll and ParquetSharpNative.so next to the assembly ? This doesn't look right...

Or should I ship 2 distinct folders and add the reference to the assembly corresponding to the os at the package loading (`__init__.py`) ?

What do you recommend ?

**Reply by filmor** [reply]
You'd have to check with the ParquetSharp devs whether this would work. Otherwise, as with all native deps, it's probably best to just distribute separate wheels.
